libminifi/src/utils/ThreadPool.cpp

```diff
@@ -0,0 +1,238 @@
+          std::unique_lock<std::mutex> lock(worker_queue_mutex_);
+          auto worker_thread = std::make_shared<WorkerThread>();
+          worker_thread->thread_ = createThread(std::bind(&ThreadPool::run_tasks, this, worker_thread));
+          if (daemon_threads_) {
+            worker_thread->thread_.detach();
+          }
+          thread_queue_.push_back(worker_thread);
+          current_workers_++;
+        }
+      }
+      {
+        std::lock_guard<std::recursive_mutex> lock(manager_mutex_);
+        std::shared_ptr<WorkerThread> thread_ref;
+        while (deceased_thread_queue_.try_dequeue(thread_ref)) {
+          std::unique_lock<std::mutex> lock(worker_queue_mutex_);
+          if (thread_ref->thread_.joinable())
+            thread_ref->thread_.join();
+          thread_queue_.erase(std::remove(thread_queue_.begin(), thread_queue_.end(), thread_ref), thread_queue_.end());
+        }
+      }
+      std::this_thread::sleep_for(waitperiod);
+    }
+  } else {
+    for (auto &thread : thread_queue_) {
+      if (thread->thread_.joinable())
+        thread->thread_.join();
+    }
+  }
+}
+
+template<typename T>
+void ThreadPool<T>::start() {
+  if (nullptr != controller_service_provider_) {
+    auto thread_man = controller_service_provider_->getControllerService("ThreadPoolManager");
+    thread_manager_ = thread_man != nullptr ? std::dynamic_pointer_cast<controllers::ThreadManagementService>(thread_man) : nullptr;
+  } else {
+    thread_manager_ = nullptr;
+  }
+
+  std::lock_guard<std::recursive_mutex> lock(manager_mutex_);
+  if (!running_) {
+    running_ = true;
+    manager_thread_ = std::move(std::thread(&ThreadPool::manageWorkers, this));
+    if (worker_queue_.size_approx() > 0) {
+      tasks_available_.notify_all();
+    }
+
+    std::lock_guard<std::mutex> quee_lock(worker_queue_mutex_);
+    delayed_scheduler_thread_ = std::thread(&ThreadPool<T>::manage_delayed_queue, this);
+  }
+}
+
+template<typename T>
+void ThreadPool<T>::stopTasks(const std::string &identifier) {
+  std::unique_lock<std::mutex> lock(worker_queue_mutex_);
+  task_status_[identifier] = false;
+}
+
+template<typename T>
+void ThreadPool<T>::shutdown() {
+  if (running_.load()) {
+    std::lock_guard<std::recursive_mutex> lock(manager_mutex_);
+    running_.store(false);
+
+    drain();
+
+    task_status_.clear();
+    if (manager_thread_.joinable()) {
+      manager_thread_.join();
+    }
+
+    delayed_task_available_.notify_all();
+    if(delayed_scheduler_thread_.joinable()) {
+      delayed_scheduler_thread_.join();
+    }
+
+    for(const auto &thread : thread_queue_){
+      if (thread->thread_.joinable())
+        thread->thread_.join();
+    }
+
+    thread_queue_.clear();
+    current_workers_ = 0;
+    while (worker_queue_.size_approx() > 0) {
+      Worker<T> task;
+      worker_queue_.try_dequeue(task);
+
+    }
+  }
+}
+
+template class utils::ThreadPool<utils::ComplexResult>;
```

Comment:
This is ugly but meant to be temporary. 

My long term goal would be to remove template type from ThreadPool. 
ComplexResult can be extended in case further decision factors pop up in the future. 
By removing the template parameter we could have one unified threadpool in MiNiFi that handles all kinds of tasks.

Reply:
We can erase the result type of the tasks after the we have served the consumers with their appropriately typed `std::future`. We shouldn't need to care about result type during execution.

Reply:
The template type of threadpool is not the result of the "computation" done by a task, but a type that can be used to determine whether the current task is to be executed again or not, and in case it is, when it should happen.

Reply:
In that case I think one type (something like what's currently called `ComplexResult`) can be used for all scheduling types and the template parameter is no longer needed.

Reply:
Completely agree, that's my long term plan.